### PR TITLE
feat(auth): add auth components

### PR DESCRIPTION
**Pull request by AGDholo**
- [x] create dialog component
- [x] create sign & login component
- [x]  create forget component
- [x] form validation
- [x] modal route support

Readme: packages/hedgehog-lab/src/components/User/Auth/readme.md

---

# Auth Component

## Support form method

You can add others into:  
`/components/User/Auth/AuthForm/IAuthFormMethods.ts`

Don't forget set the state  
`const [authAction, setAuthAction] = useRecoilState(authActionState)`

All logic fellow the state, view the form component file that you can find the code details.

## Add a new form

1. Set the method name, you should add the name into the `IAuthFormMethods`, and `setAuthAction` first.
2. Set the form validation and submit method, follow the blow file easily.
3. Finally, render the component with `BaseForm`.

## About the components

* The base form(`/components/Base/Input`, `/components/Base/Form`) which supports the basic method, no business code
  needs to be added.
* Business code pls put into the `/components/User/Auth` folder.

## About the file name

* State(Recoil): R*States
* Interface: I*(FunctionName: Props, Methods etc...)

### Form validation rules

`/components/User/Auth/AuthForm/AuthFormYup.ts`

### Form submit method

`/components/User/Auth/AuthForm/AuthFormSubmit.ts`

---

## Demo:

https://user-images.githubusercontent.com/13300332/141168348-d414b41a-8329-4aa4-8c84-ca5a7c239989.mp4

### CODE REVIEW

File: packages/hedgehog-lab/src/components/Header/Header.tsx [hunk 1]
```diff
@@ -62,7 +69,7 @@ const encodingShareableUrlWithURL = (yourUrl: string, autoRun: boolean): string
 
 
 const Header: React.FC<HeaderProps> = (props: HeaderProps) => {
-    const { siderBarOpen, setOpen, lgBreakpointMatches, source } = props;
```
Comment (Himself65) [inline packages/hedgehog-lab/src/components/Header/Header.tsx:65]: Should we add lint rules for all of these?

File: packages/hedgehog-lab/src/components/Header/Header.tsx [hunk 2]
```diff
@@ -133,66 +140,65 @@ const Header: React.FC<HeaderProps> = (props: HeaderProps) => {
...
+                            aria-label="menu"
+                            style={{display: 'inline'}}
+                            onClick={handleSideBarOpen}
+                            size="large">
+                            <MenuOutlined style={{fontSize: '1.25rem'}}/>
+                        </IconButton>
+
+                        {lgBreakpointMatches &&
+                        <Typography
+                            variant={'h6'}
```
Comment (Himself65) [inline packages/hedgehog-lab/src/components/Header/Header.tsx:159]: `variant='h6'`

File: packages/hedgehog-lab/src/components/Header/Header.tsx [hunk 3]
```diff
@@ -23,11 +21,20 @@ import TextField from '@mui/material/TextField';
...
 import FormControlLabel from '@mui/material/FormControlLabel';
 import Box from '@mui/material/Box';
-import { HEDGEHOG_DOMAIN } from "../../config"
-import { CopyAllOutlined, InsertDriveFileOutlined, MenuOutlined, NightsStayOutlined, ShareOutlined, WbSunnyOutlined } from "@mui/icons-material";
-import { useCopyToClipboard } from "react-use";
-import { useSnackbar } from "notistack";
-import { ColorModeContext } from '../../App';
+import {HEDGEHOG_DOMAIN} from "../../config"
+import {
+    CopyAllOutlined,
```
Comment (Himself65) [inline packages/hedgehog-lab/src/components/Header/Header.tsx:26]: add the suffix `Icon`, like `CopyAllOutlined` should be like `CopyAllOutlined as CopyAllOutlinedIcon`

File: packages/hedgehog-lab/src/components/User/AuthDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            Login
+        </Button>
+    )
+}
+
+const DialogHeader = (props: DialogProps) => {
+    const {handleClose} = props
+    return (
+        <DialogTitle>
+            <Typography variant={'h4'} component={'p'} sx={{fontWeight: 'bold'}}>
```
Comment (Himself65) [inline packages/hedgehog-lab/src/components/User/AuthDialog.tsx:39]: remove brackets '{}'

File: packages/hedgehog-lab/src/components/User/AuthDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            <OutlinedInput error fullWidth placeholder={'Email'} startAdornment={
+                <InputAdornment position={"start"}>
+                    <EmailOutlined/>
+                </InputAdornment>
+            }/>
+
+            <OutlinedInput
+                type={values.showPassword ? 'text' : 'password'}
+                value={values.password}
+                onChange={handleChange('password')}
```
Comment (Himself65) [inline packages/hedgehog-lab/src/components/User/AuthDialog.tsx:92]: you only use this function factor once

**Review by Himself65 — COMMENTED** [review]
just conscious, why you don't use `useCallback` for each function in react component?

File: packages/hedgehog-lab/src/components/Header/Header.tsx (hunk 1 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:65]: yeah, I agress

File: packages/hedgehog-lab/src/components/Header/Header.tsx (hunk 2 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:159]: Why not {}?, I guess two can, my IDE is automatically completed {},

File: packages/hedgehog-lab/src/components/User/AuthDialog.tsx (hunk 5 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/User/AuthDialog.tsx:92]: sorry, what's mean? I don't really understand this.

File: packages/hedgehog-lab/src/components/Header/Header.tsx (hunk 3 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:26]: yeah, this makes the semantics clearer, but it can be a bit cumbersome.

Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:26]: note: https://mui.com/components/icons/#usage

File: packages/hedgehog-lab/src/components/Header/Header.tsx (hunk 2 above)
Comment (Himself65) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:159]: it's fine, this is my code style

File: packages/hedgehog-lab/src/components/User/AuthDialog.tsx (hunk 5 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/User/AuthDialog.tsx:92]: I delete it, React Hook Form will be used later.
Call function once will be enabled in subsequent code, along with ESLint.

File: packages/hedgehog-lab/src/components/Header/Header.tsx (hunk 3 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:26]: I guess, we just need add the alias name to the confusing words, such as individual GitHub named icon.

There is a unique suffix to indicate the icon name(outlined etc...), which is clear enough.

Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:26]: If needs to modify or make other changes, we may want to present a separate issue to discuss in detail.

File: packages/hedgehog-lab/src/components/Header/Header.tsx (hunk 1 above)
Comment (AGDholo) [reply packages/hedgehog-lab/src/components/Header/Header.tsx:65]: Detailed changes will be made later in other PR.

**Pull request merged**